### DISCUSSION TITLE
(Question) Deprecation of SDL1

### DISCUSSION THREAD

**Discussion by dreamer** [ihhub/fheroes2#3916] [General]
Hello, just discovered this project via /r/linux_gaming and I must admit - I'm impressed :) It seems like you're on a good track to proper reimplementation/modernization, congratulations to the team!

When I'll have some spare cycles, I would like to contribute as well, I just have a few questions (I haven't found any reference to any Discord, IRC, or Matrix channel, otherwise I would ask directly in there).

Do you have a plan to retiring SDL1 support? I'm asking, because I know there are some quite non-obvious differences between SDL1 and SDL2 - developing project while managing compatibility with both seems like a waste of precious resources (testing time, development time). By having a quick look at your bugtracker I noticed several places where I could help out, but I would prefer to keep things simple and stick to SDL2 only for new code.

**Comment by ihhub** [comment]
Hi @dreamer , thanks for interest!

We do not have plans to retire SDL1 as of now because SDL2 still has more issues than SDL1. In the future we're going to stick to one library which is SDL2. So in short, yes, we are going but not right now :)

**Comment by dreamer** [comment]
Roger that :) Sensible design/maintenance choice - from my experience I can attest, that moving to SDL2 results in fewer problems overall (especially on newer systems), and new APIs help with maintenance and implementation of new features.

Follow up questions (more technical now):

- Are you more interested in moving codebase to SDL2 or to C++11 at the moment? (I noticed #254 - I think it can be somewhat easily fixed via a semantic patch to upgrade this usage straight to lambda expressions).
- By briefly looking at the code, I see you're using SDL_Surface for drawing - is this the only method of pushing pixels? Update to SDL2 SDL_Texture could make it easy to provide initial implementation of #286, or maybe even attaching GLSL OpenGL shaders for nice upscaling of the game window (shader implementations can be largely copied from [the project I'm mostly working on ATM](https://github.com/dosbox-staging/dosbox-staging)).
- Documentation and scripts did not refer to SDL 1.2 library SDL_cdrom (and code compiled successfully), but I noticed some code, that seems to use it. Do you plan to support playing music from physical CD in the future or is it not a priority at all? (for my project I decided to drop it completely - SDL2 does not provide this library and users don't really care about physical CD support any more).
- Are you interested in putting some additional gating on CI? (in the project I linked we have scripts for counting warnings, that fail CI builds if PR introduces new warnings in the code, I think it would be quite easy to port them to this project)

**Comment by ihhub** [comment]
Hi @dreamer ,

> * Are you more interested in moving codebase to SDL2 or to C++11 at the moment? (I noticed #254 - I think it can be somewhat easily fixed via a semantic patch to upgrade this usage straight to lambda expressions).
- moving to **C++11** is a higher priority for now. The original code was written in **C++03** and as you have noticed we need to fix some places to be able to do this. Moving to SDL2 is a future goal with dropping a support for SDL 1 (as I explained in the comment above)
> * By briefly looking at the code, I see you're using SDL_Surface for drawing - is this the only method of pushing pixels? Update to SDL2 SDL_Texture could make it easy to provide initial implementation of #286, or maybe even attaching GLSL OpenGL shaders for nice upscaling of the game window (shader implementations can be largely copied from [the project I'm mostly working on ATM](https://github.com/dosbox-staging/dosbox-staging)).
- this is the only way to draw for now. We could use other ways for drawings but we need to keep in mind portability of every solution. We have an API within the project for scaling: `Surface::RenderScale`
> * Documentation and scripts did not refer to SDL 1.2 library SDL_cdrom (and code compiled successfully), but I noticed some code, that seems to use it. Do you plan to support playing music from physical CD in the future or is it not a priority at all? (for my project I decided to drop it completely - SDL2 does not provide this library and users don't really care about physical CD support any more).
- this is not a priority for now and most likely we need to drop it in the future
> * Are you interested in putting some additional gating on CI? (in the project I linked we have scripts for counting warnings, that fail CI builds if PR introduces new warnings in the code, I think it would be quite easy to port them to this project)
- yes but keep in mind that a huge number of warnings still present in the code. We have to fix them. Many of the warnings have been fixed in the past. We have **AppVeyor** and **Travis** CI/CD pipelines, we could integrate a script there.

**Comment by egorpugin** [comment]
Moving to C++17/20 will boost your codebase in terms of available modern features.

**Comment by ihhub** [comment]
Hi @egorpugin , could you please elaborate how it'll boost the codebase?

**Comment by egorpugin** [comment]
I've opened some random file:

Write `auto` in lambdas instead of types (c++14?)
https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/kingdom/kingdom.cpp#L139

range for loop (c++11?)
https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/kingdom/kingdom.cpp#L171
ranges? (c++20)
```
for ( EventsDate::const_iterator it = events.begin(); it != events.end(); ++it )
        AddFundsResource( ( *it ).resource );
```
goes into
```
std::ranges::for_each(events, AddFundsResource, &/*that even type*/::resource);
```
(example - https://godbolt.org/z/1qP3o8je6)

---

What you get in general:
- it is easier to write code
- it is easier to maintain code
- you can write more generic code
- you write shorter code
- it is more efficient code
- it is less error prone code
- codebase lives

Some other considerations:
- gcc 11 will make C++17 the default mode (currently C++14).
- google recommends C++17 (https://google.github.io/styleguide/cppguide.html#C++_Version)
- more...?

---

If you have any specific concerns, tell them, maybe I could help somehow.

upd.:
You don't need to rewrite old code into modern. But when writing fresh code, new features will be available.
People may send PRs faster having actual C++ standard active.
etc.

**Comment by ihhub** [comment]
Hi @egorpugin , we actually had a discussion about this and there's no solid advantage of doing it except complicating code as not every developer know all new features. So easy is a subjective opinion🙂
We moved from 03 to 11 to have move semantics, future multithreading support and to use lambdas. Using auto leads to many issues such as developers tend to forget to pass by reference rather than by value.
If 17/20 format would really bring benefit for the project then it'll be worth to switch (keeping in mind old compilers).

As for highlighted example: we could use **for (const auto & event : events )** which is supported by C++11.
